components/entities/api/src/main/java/org/phenotips/entities/internal/DefaultPrimaryEntityResolver.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.entities.internal;
+
+import org.phenotips.entities.PrimaryEntity;
+import org.phenotips.entities.PrimaryEntityManager;
+import org.phenotips.entities.PrimaryEntityResolver;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Default implementation of the {@link PrimaryEntityResolver} component, which uses all the
+ * {@link PrimaryEntityManager entity managers} registered in the component manager.
+ *
+ * @version $Id$
+ * @since 1.4
+ */
+@Component
+@Singleton
+public class DefaultPrimaryEntityResolver implements PrimaryEntityResolver, Initializable
+{
+    /** The currently available primary entity managers. */
+    @Inject
+    private Map<String, PrimaryEntityManager> repositories;
+
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> referenceParser;
+
+    /** Currently available primary entity managers, mapped by their entity ID prefix. */
+    private Map<String, PrimaryEntityManager> repositoriesByPrefix;
+
+    /** Currently available primary entity managers, mapped by their type. */
+    private Map<String, PrimaryEntityManager> repositoriesByType;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        try {
+            this.repositoriesByPrefix = new HashMap<>();
+            this.repositoriesByType = new HashMap<>();
+
+            this.repositories.values().forEach(this::addToMaps);
+        } catch (final RuntimeException ex) {
+            throw new InitializationException(ex.getMessage());
+        }
+    }
+
+    @Nullable
+    @Override
+    public PrimaryEntity resolveEntity(@Nullable final String entityId)
```

Comment:
```
do arguments sanity check
extract id prefix
for each manager
  if manager's prefix is the same as the entity prefix
    get the primary entity from it
    if not null
      return
return null
```

Not sure about checking the prefix, though, we could just try to resolve the entity, but that will affect performance slightly. For example, on PhenomeCentral we have several patients starting with `F` (very old patients from before PhenomeCentral actually existed, and the `F` stands for `Forge`, the project that later became `Care4Rare`). Maybe we could do a first round with prefix checking, and if nothing is found, do another round without prefix checking.

Reply:
Fixed.